src/mlpack/tests/rl_components_test.cpp

```diff
@@ -227,7 +227,7 @@ BOOST_AUTO_TEST_CASE(RandomReplayTest)
   double reward = env.Sample(state, action, nextState);
   replay.Store(state, action, reward, nextState, env.IsTerminal(nextState));
   arma::mat sampledState;
-  arma::icolvec sampledAction;
+  std::vector<MountainCar::Action> sampledAction;
```

Comment:
Can you test PPO on the ``DoublePoleCart`` env, see https://github.com/mlpack/mlpack/blob/12a50a055ba7f69340598329bd146ee37bec110f/src/mlpack/tests/q_learning_test.cpp#L401 for an example as well. Compared with the ``MountainCar`` env it's a more challenging to solve the env.

Reply:
> Can you test PPO on the `DoublePoleCart` env, see
> 
> https://github.com/mlpack/mlpack/blob/12a50a055ba7f69340598329bd146ee37bec110f/src/mlpack/tests/q_learning_test.cpp#L401
> 
> for an example as well. Compared with the `MountainCar` env it's a more challenging to solve the env.

Yes, I know this task. But currently, the PPO implement of mine is only supporting continuous task like `PENDULUM`.

Reply:
Should we implement another continuous task, that is more challenging, e.g. what do you think about LunarLander: https://github.com/zippkidd/lunarLander?

Reply:
> Should we implement another continuous task, that is more challenging, e.g. what do you think about LunarLander: https://github.com/zippkidd/lunarLander?

Yes, LunarLander is another continuous task, it is more challenging than Pendulum. But I am not sure how many workload of that, I can try to do that.

Reply:
You can take a look at the repo I referenced above, there are mainly two functions that are interesting:

```c++
void updateStatus(double &velocity, double burnAmount, double &fuelRemaining, double &height)
{
    fuelRemaining = fuelRemaining - burnAmount;
    height = height - ((((velocity - burnAmount) + 5) + velocity)/2);
    velocity = velocity - burnAmount + 5;
}
void touchdown(double &elapsedTime, double &velocity, double &burnAmount, double &height)
{
    double oldvelocity = velocity + burnAmount - 5;
    double oldheight = height + ((velocity + oldvelocity)/2);
    double oldtime = elapsedTime - 1;
    double fraction = 0.0;
    if (burnAmount == 5)
        fraction = oldheight / oldvelocity;
    else
        fraction = (sqrt((oldvelocity*oldvelocity) + (oldheight*(10 - (2 * burnAmount)))) - oldvelocity) / (5 - burnAmount);
    elapsedTime = oldtime + fraction*1.0;
    velocity = oldvelocity + (5 - burnAmount)*fraction *1.0;
    height = oldheight;
}
```

Reply:
> You can take a look at the repo I referenced above, there are mainly two functions that are interesting:
> 
> ```c++
> void updateStatus(double &velocity, double burnAmount, double &fuelRemaining, double &height)
> {
>     fuelRemaining = fuelRemaining - burnAmount;
>     height = height - ((((velocity - burnAmount) + 5) + velocity)/2);
>     velocity = velocity - burnAmount + 5;
> }
> void touchdown(double &elapsedTime, double &velocity, double &burnAmount, double &height)
> {
>     double oldvelocity = velocity + burnAmount - 5;
>     double oldheight = height + ((velocity + oldvelocity)/2);
>     double oldtime = elapsedTime - 1;
>     double fraction = 0.0;
>     if (burnAmount == 5)
>         fraction = oldheight / oldvelocity;
>     else
>         fraction = (sqrt((oldvelocity*oldvelocity) + (oldheight*(10 - (2 * burnAmount)))) - oldvelocity) / (5 - burnAmount);
>     elapsedTime = oldtime + fraction*1.0;
>     velocity = oldvelocity + (5 - burnAmount)*fraction *1.0;
>     height = oldheight;
> }
> ```

Oh, yes, I read that implementation, it is not so complicated. But I found that it is different from the Lunarlander game in the GYM.
The gym implementation's action is [Action is two real values vector from -1 to +1. First controls main engine, -1..0 off, 0..+1 throttle from 50% to 100% power. The engine can't work with less than 50% power. Second value -1.0..-0.5 fire left engine, +0.5..+1.0 fire right engine, -0.5..0.5 off.](https://github.com/openai/gym/wiki/Leaderboard#lunarlandercontinuous-v2)

In this [implementation](https://github.com/zippkidd/lunarLander/blob/master/LanderSim.cpp), the action is the amount of fuel to use.

Reply:
What do you think of the [continuous mountain car task](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/reinforcement_learning/environment/continuous_mountain_car.hpp)?

The implementation in [gym.](https://github.com/openai/gym/blob/master/gym/envs/classic_control/continuous_mountain_car.py)

Reply:
The MountainCar env is similar to the Pendulum env, we could test against the ``ContinuousDoublePoleCart`` which is more difficult, not as difficult as the lunar lander env.

> Oh, yes, I read that implementation, it is not so complicated. But I found that it is different from the Lunarlander game in the GYM.
> The gym implementation's action is Action is two real values vector from -1 to +1. First controls main engine, -1..0 off, 0..+1 throttle from 50% to 100% power. The engine can't work with less than 50% power. Second value -1.0..-0.5 fire left engine, +0.5..+1.0 fire right engine, -0.5..0.5 off.
> 
> In this implementation, the action is the amount of fuel to use.
> 

We could go with the OpenAI gym settings.

Reply:
> The MountainCar env is similar to the Pendulum env, we could test against the `ContinuousDoublePoleCart` which is more difficult, not as difficult as the lunar lander env.
> 
> > Oh, yes, I read that implementation, it is not so complicated. But I found that it is different from the Lunarlander game in the GYM.
> > The gym implementation's action is Action is two real values vector from -1 to +1. First controls main engine, -1..0 off, 0..+1 throttle from 50% to 100% power. The engine can't work with less than 50% power. Second value -1.0..-0.5 fire left engine, +0.5..+1.0 fire right engine, -0.5..0.5 off.
> > In this implementation, the action is the amount of fuel to use.
> 
> We could go with the OpenAI gym settings.

Can we implement the env use the setting in [there](https://github.com/zippkidd/lunarLander/blob/master/LanderSim.cpp), then go with the OpenAI gym settings step by step?

Reply:
Sounds reasonable, the basic idea is to see if PPO works on more challenging tasks as well as for the current Pendulum env.

Reply:
Let me know if you need any help with the implementation.

Reply:
> Let me know if you need any help with the implementation.

Oh, I am implementing the Lunarlander now, can I implement it in this PR?

Reply:
> > Let me know if you need any help with the implementation.
> 
> Oh, I am implementing the Lunarlander now, can I implement it in this PR?

I can not figure out why one of the CI failed.

Reply:
> > > Let me know if you need any help with the implementation.
> > 
> > 
> > Oh, I am implementing the Lunarlander now, can I implement it in this PR?
> 
> I can not figure out why one of the CI failed, can you take a look?

Reply:
> Oh, I am implementing the Lunarlander now, can I implement it in this PR?

I think that would be the best idea.

Reply:
The appveyor build failure isn't related with the changes from the PR, looks like the latest MSVC version has some issue with loading arff files.

Reply:
With the implementation in [here](https://github.com/zippkidd/lunarLander/blob/master/LanderSim.cpp), I found that we hardly determine the reward of the agent. What's more, I can not define when the agent is solved. 

In the [gym's setting](https://github.com/openai/gym/wiki/Leaderboard), the Lunarlander defines "solving" as getting an average reward of 200 over 100 consecutive trials. 

So I think we can go with the OpenAI's setting was better, what do you think of it? Can you give me some time to figure out the gym's implementation?

Reply:
> In the [gym's setting](https://github.com/openai/gym/wiki/Leaderboard), the Lunarlander defines "solving" as getting an average reward of 200 over 100 consecutive trials.

The implementation always implements a single trial, so the user has to run the same model against the env 100 consecutive trials in a row (starting with a need random starting point) to check if the model is able to solve the env.

But you are right a single Lunarlander iteration is solved if you get 200 points.

> So I think we can go with the OpenAI's setting was better, what do you think of it? Can you give me some time to figure out the gym's implementation?

Sure, let me know if you need any help.